### PR TITLE
feat: V2 Org Memberships

### PR DESCRIPTION
**Pull request by alishaz-polymath**
## What does this PR do?

`/organizations/:orgId/membership` - isOrganizationGuard

GET - RolesGuard (ADMIN) - Get all memberships within the organization
POST - RolesGuard (ADMIN) - Create a membership within the organization

`/organizations/:orgId/membership/:membershipId` - isOrganizationGuard

GET - RolesGuard (ADMIN) - Get a membership within the organization
PATCH - RolesGuard (ADMIN) - Can only patch a membership that is part of the organization
DELETE - RolesGuard (ADMIN) - Can only delete a membership already part of the organization

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [x] N/A - docs are auto generated, I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [x] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

## How should this be tested?

yarn test:e2e organizations-membership.controller.e2e-spec.ts

### CODE REVIEW

File: apps/api/v2/src/modules/organizations/controllers/organizations-membership.controller.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+export class OrganizationsMembershipsController {
+  constructor(private organizationsMembershipService: OrganizationsMembershipService) {}
+
+  @Get()
+  @UseGuards()
+  @Roles("ORG_ADMIN")
+  async getAllMemberships(
+    @Param("orgId", ParseIntPipe) orgId: number,
+    @Query() queryParams: SkipTakePagination
+  ): Promise<ApiResponse<OrgMembershipOutputDto[]>> {
```
Comment (supalarry) [inline apps/api/v2/src/modules/organizations/controllers/organizations-membership.controller.ts:45]: You have defined return type of the routes as :
```
Promise<ApiResponse<OrgMembershipOutputDto[]>>
```

but this would not generate swagger response, because ApiResponse is just a type, and not a class-validator class. For example see `v2/src/ee/schedules/schedules_2024_04_15/controllers/schedules.controller.ts` - it returns `Promise<CreateScheduleOutput_2024_04_15>`, which looks like:

```
export class CreateScheduleOutput_2024_04_15 {
  @ApiProperty({ example: SUCCESS_STATUS, enum: [SUCCESS_STATUS, ERROR_STATUS] })
  @IsEnum([SUCCESS_STATUS, ERROR_STATUS])
  status!: typeof SUCCESS_STATUS | typeof ERROR_STATUS;

  @ApiProperty({
    type: ScheduleOutput,
  })
  @IsNotEmptyObject()
  @ValidateNested()
  @Type(() => ScheduleOutput)
  data!: ScheduleOutput;
}

``` 

this makes sure that the swagger doc generated has status and data keys.

Comment (supalarry) [reply apps/api/v2/src/modules/organizations/controllers/organizations-membership.controller.ts:45]: To check the swagger doc generated start the api and go to http://localhost:5555/docs

File: apps/api/v2/src/modules/organizations/inputs/create-organization-membership.input.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+export class CreateOrgMembershipDto {
+  @IsInt()
+  readonly teamId!: number;
+
+  @IsInt()
+  readonly userId!: number;
+
+  @IsOptional()
+  @IsBoolean()
+  readonly accepted?: boolean = false;
```
Comment (supalarry) [inline apps/api/v2/src/modules/organizations/inputs/create-organization-membership.input.ts:14]: Following up on [campsite comment](https://app.campsite.co/cal/posts/rzfttn7s2cm0#comment-oupbuqm9zi53)  - my first instint is that we should not allow passing accepted at all, because I think the users should manually accept the membership.

File: apps/api/v2/src/modules/organizations/inputs/create-organization-membership.input.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+  @IsString()
```
Comment (supalarry) [inline apps/api/v2/src/modules/organizations/inputs/create-organization-membership.input.ts:16]: Even though the imported "MembershipRole" is defined as const instead of enum, I think we can use @IsEnum(MembershipRole) - i would add this and then try making requests with correct and incorrect membersip role and see if it blocks incorrect roles, cuz right now any string can be passed.

File: apps/api/v2/src/modules/organizations/repositories/organizations-membership.repository.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
+import { CreateOrgMembershipDto } from "@/modules/organizations/inputs/create-organization-membership.input";
+import { PrismaReadService } from "@/modules/prisma/prisma-read.service";
+import { PrismaWriteService } from "@/modules/prisma/prisma-write.service";
+import { Injectable } from "@nestjs/common";
+
+@Injectable()
+export class OrganizationsMembershipRepository {
+  constructor(private readonly dbRead: PrismaReadService, private readonly dbWrite: PrismaWriteService) {}
+
+  private async findAllTeamsInOrg(organizationId: number) {
```
Comment (supalarry) [inline apps/api/v2/src/modules/organizations/repositories/organizations-membership.repository.ts:10]: I think this belongs in `organizations-teams.repository.ts`

Comment (supalarry) [reply apps/api/v2/src/modules/organizations/repositories/organizations-membership.repository.ts:10]: I see `findAllTeamsInOrg` is used in this repository below - I would dependency inject `organizations-teams.repository.ts` into the organizations membersip service, which gets IDs of all teams. And then they are passed as parameters to this repository function instead of organizationId e.g.:

```
 async findOrgMembershipsPaginated(organizationTeamsIds: number[], skip: number, take: number) {
    return this.dbRead.prisma.membership.findMany({
      where: {
        teamId: { in: organizationTeamsIds },
      },
      skip,
      take,
    });
  }
```

**Review by supalarry — COMMENTED** [review]
Great work, have some refactor comments below - let me know what you think.

File: apps/api/v2/src/modules/organizations/inputs/create-organization-membership.input.ts (hunk 2 above)
Comment (alishaz-polymath) [reply apps/api/v2/src/modules/organizations/inputs/create-organization-membership.input.ts:14]: Let's discuss in campsite further on the approach here 🙏

**Comment by alishaz-polymath** [comment]
> Great work, have some refactor comments below - let me know what you think.

Awesome review 🙏 
Will start taking them down

File: apps/api/v2/src/modules/organizations/controllers/organizations-membership.controller.ts (hunk 1 above)
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/organizations/controllers/organizations-membership.controller.ts:45]: I would avoid putting a suffix version except if we really need to differentiate with the base version

File: apps/api/v2/src/modules/organizations/controllers/organizations-membership.controller.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+@Controller({
+  path: "/v2/organizations/:orgId/memberships",
+  version: API_VERSIONS_VALUES,
+})
+@UseGuards(ApiAuthGuard, IsOrgGuard, RolesGuard)
+@DocsTags("Organizations Memberships")
+export class OrganizationsMembershipsController {
+  constructor(private organizationsMembershipService: OrganizationsMembershipService) {}
+
+  @Get()
```
[the remark this answers is not in the document]
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/organizations/controllers/organizations-membership.controller.ts:39]: yes it defaults to "/" but we can indeed add it for consistency

File: apps/api/v2/src/modules/organizations/inputs/create-organization-membership.input.ts (hunk 2 above)
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/organizations/inputs/create-organization-membership.input.ts:14]: I think when an admin is creating a membership directly he should be able to do whatever

for the /invite endpoint it's a different story and accepted should not be passed

File: apps/api/v2/src/modules/organizations/repositories/organizations-teams.repository.ts [hunk 6]
```diff
@@ -55,4 +55,24 @@ export class OrganizationsTeamsRepository {
...
+
+  async findOrgTeams(organizationId: number) {
+    const teams = await this.dbRead.prisma.team.findMany({
+      where: {
+        OR: [
+          {
+            parentId: organizationId,
+          },
+          {
+            id: organizationId,
```
Comment (supalarry) [inline apps/api/v2/src/modules/organizations/repositories/organizations-teams.repository.ts:67]: Do we want to put organization itself and teams in the same box? From the business perspective I feel like I would exclude the `id: organizationId` rule because orgs and teams are perceived differently, and instead we could have a function `findOrg` if we needed one.

File: apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
+import { API_VERSIONS_VALUES } from "@/lib/api-versions";
```
Comment (supalarry) [inline apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts:1]: Would be good to add e2e tests to the controller - we try to have each controller covered. For example can see `controllers/organizations-teams.controller.e2e-spec.ts`.

**Review by supalarry — COMMENTED** [review]
Great work! Looks good with 2 minor NITs, but what is missing are e2e tests.

File: apps/api/v2/src/modules/organizations/repositories/organizations-teams.repository.ts [hunk 8]
```diff
@@ -55,4 +55,24 @@ export class OrganizationsTeamsRepository {
...
+          },
+        ],
+      },
+      select: {
+        id: true,
+      },
+    });
+
+    return teams.map((team) => team.id);
+  }
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/repositories/organizations-teams.repository.ts:77]: this function is not necessary, check other comments

File: apps/api/v2/src/modules/organizations/repositories/organizations-membership.repository.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  async findOrgMembershipsPaginated(organizationTeamsIds: number[], skip: number, take: number) {
+    return this.dbRead.prisma.membership.findMany({
+      where: {
+        teamId: { in: organizationTeamsIds },
+      },
+      skip,
+      take,
+    });
+  }
+}
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/repositories/organizations-membership.repository.ts:50]: findOrgTeams is not necessary, findOrgMemberships is currently get the memberships of the teams of an org, not the memberships of the org, which is confusing, we should make another PR with another controller to handle memberships of the teams within an org

```ts

 async findOrgMembershipsPaginated(organizationId: number, skip: number, take: number) {
    return this.dbRead.prisma.membership.findMany({
      where: {
       teamId: organizationId,
      },
      skip,
      take,
    });
  }
  
  ```

File: apps/api/v2/src/modules/organizations/repositories/organizations-teams.repository.ts (hunk 8 above)
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/organizations/repositories/organizations-teams.repository.ts:77]: we need another PR to handle memberships of teams within orgs

File: apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    return {
+      status: SUCCESS_STATUS,
+      data: plainToClass(OrgMembershipOutputDto, membership, { strategy: "excludeAll" }),
+    };
+  }
+
+  @Roles("ORG_ADMIN")
+  @UseGuards(IsMembershipInOrg)
+  @Delete("/:membershipId")
+  @HttpCode(HttpStatus.OK)
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts:95]: when returning 200, we should return the deleted entity

if we don't want to return content use HttpStatus.NO_CONTENT

File: apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    };
+  }
+
+  @UseGuards(IsMembershipInOrg)
+  @Roles("ORG_ADMIN")
+  @Patch("/:membershipId")
+  async updateMembership(
+    @Param("orgId", ParseIntPipe) orgId: number,
+    @Param("membershipId", ParseIntPipe) membershipId: number,
+    @Body() body: CreateOrgMembershipDto
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts:112]: we should be able to partially update membership

```@Body() body: Partial<CreateOrgMembershipDto>```

File: apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  @Roles("ORG_ADMIN")
+  @UseGuards(IsMembershipInOrg)
+  @Get("/:membershipId")
+  async getUserSchedule(
+    @Param("orgId", ParseIntPipe) orgId: number,
+    @Param("membershipId", ParseIntPipe) membershipId: number
+  ): Promise<GetOrgMembership> {
+    const membership = await this.organizationsMembershipService.getOrgMembership(orgId, membershipId);
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts:84]: the guard IsMembershipInOrg is already populating the request with the correct membership, use the decorator      @GetMembership instead

File: apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+      data: plainToClass(OrgMembershipOutputDto, membership, { strategy: "excludeAll" }),
+    };
+  }
+
+  @Roles("ORG_ADMIN")
+  @UseGuards(IsMembershipInOrg)
+  @Get("/:membershipId")
+  async getUserSchedule(
+    @Param("orgId", ParseIntPipe) orgId: number,
+    @Param("membershipId", ParseIntPipe) membershipId: number
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts:82]: ```@GetMembership() membership: Membership```

File: apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts (hunk 10 above)
Comment (alishaz-polymath) [reply apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts:95]: About the status codes, wouldn't it make sense to return 204 for deleted intead of 200?

File: apps/api/v2/src/modules/auth/guards/memberships/is-membership-in-org.guard.ts [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+      Number(orgId),
+      Number(membershipId)
+    );
+
+    if (membership) {
+      request.membership = membership;
+      return true;
+    }
+
+    if (!membership) {
```
[the remark this answers is not in the document]
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/auth/guards/memberships/is-membership-in-org.guard.ts:40]: I agree

File: apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts (hunk 10 above)
Comment (alishaz-polymath) [reply apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts:95]: Also 201 for Create

Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts:96]: 204 is returned by using HttpStatus.NO_CONTENT yes

Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/organizations/controllers/memberships/organizations-membership.controller.ts:96]: > Also 201 for Create

yes indeed, let's start using HttpStatus.CREATED for that

**Review by ThyMinimalDev — DISMISSED** [review]
Looking good, followup pr for memberships of teams within an org

**Review by ThyMinimalDev — APPROVED** [review]
